lib/elixir/lib/module/parallel_checker.ex

```diff
@@ -0,0 +1,268 @@
+    preload_cache(ets, modules)
+
+    state = %{
+      ets: ets,
+      waiting: %{},
+      reply_when_finished: nil,
+      modules: modules,
+      spawned: 0,
+      schedulers: schedulers,
+      binaries: [],
+      warnings: []
+    }
+
+    {:ok, state}
+  end
+
+  def handle_call(:start_and_wait, from, %{reply_when_finished: nil} = state) do
+    state = spawn_checkers(%{state | reply_when_finished: from})
+    {:noreply, state}
+  end
+
+  def handle_call({:lock, module}, from, %{waiting: waiting} = state) do
+    case waiting do
+      %{^module => froms} ->
+        waiting = :maps.put(module, [from | froms], state.waiting)
+        {:noreply, %{state | waiting: waiting}}
+
+      %{} ->
+        waiting = :maps.put(module, [], state.waiting)
+        {:reply, true, %{state | waiting: waiting}}
+    end
+  end
+
+  def handle_call({:unlock, module}, _from, %{waiting: waiting} = state) do
+    froms = :maps.get(module, waiting)
+    Enum.each(froms, &:gen_server.reply(&1, false))
+    waiting = :maps.remove(module, waiting)
+    {:reply, :ok, %{state | waiting: waiting}}
+  end
+
+  def handle_info({__MODULE__, module, binary, warnings}, state) do
+    state = %{
+      state
+      | binaries: [{module, binary} | state.binaries],
+        warnings: warnings ++ state.warnings,
+        spawned: state.spawned - 1
+    }
+
+    if state.spawned == 0 and state.modules == [] and not state.reply_when_finished do
+      :gen_server.reply(state.reply_when_finished, {state.binaries, state.warnings})
+      {:noreply, state}
+    else
+      state = spawn_checkers(state)
+      {:noreply, state}
+    end
+  end
+
+  defp start_and_wait(server) do
+    :gen_server.call(server, :start_and_wait, :infinity)
+  end
+
+  defp lock(server, module) do
+    :gen_server.call(server, {:lock, module}, :infinity)
+  end
+
+  defp unlock(server, module) do
+    :gen_server.call(server, {:unlock, module})
+  end
+
+  defp preload_cache(ets, modules) do
+    Enum.each(modules, fn {map, _binary} ->
+      exports = [{{:__info__, 1}, :def} | definitions_to_exports(map.definitions)]
+      deprecated = :maps.from_list(map.deprecated)
+      cache_module(ets, map.module, exports, deprecated)
+    end)
+  end
+
+  defp spawn_checkers(%{modules: []} = state) do
+    state
+  end
+
+  defp spawn_checkers(%{spawned: spawned, schedulers: schedulers} = state)
+       when spawned >= schedulers do
+    state
+  end
+
+  defp spawn_checkers(%{modules: [{map, binary} | modules]} = state) do
+    parent = self()
+    ets = state.ets
+
+    spawn_link(fn ->
+      {binary, warnings} = Module.Checker.verify(map, binary, {parent, ets})
+      send(parent, {__MODULE__, map.module, binary, warnings})
+    end)
+
+    spawn_checkers(%{state | modules: modules, spawned: state.spawned + 1})
+  end
+
+  defp cache_module({server, ets}, module) do
+    if lock(server, module) do
+      cache_from_chunk(ets, module) || cache_from_info(ets, module)
```

Comment:
I think this would be clearer with something like:

```elixir
case fetch_from_chunk_or_info(module) do 
  {:ok, exports, deprecated} -> cache_module(ets, module, exports, deprecated)
  :error -> :ets.insert(ets, {{:cached, module}, false})
end
```

Reply:
We can't do that anymore due to how Jose suggested storing in the chunk. We need separate implementation for caching from __info__ and the chunk now.